File: examples/src/test/scala/docs/cassandra/WordCountDocExample.scala
/*
 * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.cassandra

import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import akka.Done
import akka.projection.ProjectionId
import akka.projection.scaladsl.Handler
import akka.projection.scaladsl.SourceProvider
import akka.stream.alpakka.cassandra.scaladsl.CassandraSession
import akka.stream.scaladsl.Source
import org.slf4j.LoggerFactory

object WordCountDocExample {

  //#envelope
  type Word = String
  type Count = Int

  final case class WordEnvelope(offset: Long, word: Word)

  //#envelope

  //#repository
  trait WordCountRepository {
    def load(id: String, word: Word): Future[Count]
    def loadAll(id: String): Future[Map[Word, Count]]
    def save(id: String, word: Word, count: Count): Future[Done]
  }
  //#repository

  class CassandraWordCountRepository(session: CassandraSession)(implicit val ec: ExecutionContext)
      extends WordCountRepository {
    val keyspace = "test"
    val table = "wordcount"

    override def load(id: String, word: Word): Future[Count] = {
      session.selectOne(s"SELECT count FROM $keyspace.$table WHERE id = ? and word = ?", id, word).map {
        case Some(row) => row.getInt("count")
        case None      => 0
      }
    }

    override def loadAll(id: String): Future[Map[Word, Count]] = {
      session.selectAll(s"SELECT word, count FROM $keyspace.$table WHERE id = ?", id).map { rows =>
        rows.map(row => row.getString("word") -> row.getInt("count")).toMap
      }
    }

    override def save(id: String, word: Word, count: Count): Future[Done] = {
      session.executeWrite(
        s"INSERT INTO $keyspace.$table (id, word, count) VALUES (?, ?, ?)",
        id,
        word,
        count: java.lang.Integer)
    }

    def createKeyspaceAndTable(): Future[Done] = {
      session
        .executeDDL(
          s"CREATE KEYSPACE IF NOT EXISTS $keyspace WITH REPLICATION = { 'class' : 'SimpleStrategy','replication_factor':1 }")
        .flatMap(_ => session.executeDDL(s"""
        |CREATE TABLE IF NOT EXISTS $keyspace.$table (
        |  id text,
        |  word text,
        |  count int,
        |  PRIMARY KEY (id, word))
        """.stripMargin.trim))
    }
  }

  //#sourceProvider
  class WordSource(implicit ec: ExecutionContext) extends SourceProvider[Long, WordEnvelope] {

    private val src = Source(
      List(WordEnvelope(1L, "abc"), WordEnvelope(2L, "def"), WordEnvelope(3L, "ghi"), WordEnvelope(4L, "abc")))

    override def source(offset: () => Future[Option[Long]]): Future[Source[WordEnvelope, _]] = {
      offset().map {
        case Some(o) => src.dropWhile(_.offset <= o)
        case _       => src
      }
    }

    override def extractOffset(env: WordEnvelope): Long = env.offset
  }
  //#sourceProvider

  object IllustrateVariables {
    //#mutableState
    class WordCountHandler extends Handler[WordEnvelope] {
      private val logger = LoggerFactory.getLogger(getClass)
      private var state: Map[Word, Count] = Map.empty

      override def process(envelope: WordEnvelope): Future[Done] = {
        val word = envelope.word
        val newCount = state.getOrElse(word, 0) + 1
        logger.info("Word count for {} is {}", word, newCount)
        state = state.updated(word, newCount)
        Future.successful(Done)
      }
    }
    //#mutableState
  }

  object IllustrateStatefulHandlerLoadingInitialState {

    //#loadingInitialState
    import akka.projection.scaladsl.StatefulHandler

    class WordCountHandler(projectionId: ProjectionId, repository: WordCountRepository)(implicit ec: ExecutionContext)
        extends StatefulHandler[Map[Word, Count], WordEnvelope] {

      override def initialState(): Future[Map[Word, Count]] = repository.loadAll(projectionId.id)

      override def process(state: Map[Word, Count], envelope: WordEnvelope): Future[Map[Word, Count]] = {
        val word = envelope.word
        val newCount = state.getOrElse(word, 0) + 1
        val newState = for {
          _ <- repository.save(projectionId.id, word, newCount)
        } yield state.updated(word, newCount)

        newState
      }
    }
    //#loadingInitialState
  }

  object IllustrateStatefulHandlerLoadingStateOnDemand {

    //#loadingOnDemand
    import akka.projection.scaladsl.StatefulHandler

    class WordCountHandler(projectionId: ProjectionId, repository: WordCountRepository)(implicit ec: ExecutionContext)
        extends StatefulHandler[Map[Word, Count], WordEnvelope] {

      override def initialState(): Future[Map[Word, Count]] =
        Future.successful(Map.empty)

      override def process(state: Map[Word, Count], envelope: WordEnvelope): Future[Map[Word, Count]] = {
        val word = envelope.word

        val currentCount =
          state.get(word) match {
            case None =>
              repository.load(projectionId.id, word)
            case Some(count) =>
              Future.successful(count)
          }

        val newState = for {
          c <- currentCount
          newCount = c + 1
          _ <- repository.save(projectionId.id, word, newCount)
        } yield state.updated(word, newCount)

        newState
      }

    }
    //#loadingOnDemand
  }

}

File: examples/src/test/java/jdocs/cassandra/WordCountDocExample.java
/*
 * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.cassandra;

//#StatefulHandler-imports
import akka.projection.javadsl.StatefulHandler;

//#StatefulHandler-imports

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import akka.Done;
import akka.NotUsed;
import akka.projection.ProjectionId;
import akka.projection.javadsl.Handler;
import akka.projection.javadsl.SourceProvider;
import akka.stream.alpakka.cassandra.javadsl.CassandraSession;
import akka.stream.javadsl.Source;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public interface WordCountDocExample {
  //#todo
  // TDOO
  //#todo

  //#envelope
  public class WordEnvelope {
    public final Long offset;
    public final String word;

    public WordEnvelope(Long offset, String word) {
      this.offset = offset;
      this.word = word;
    }
  }

  //#envelope

  //#repository
  public interface WordCountRepository {
    CompletionStage<Integer> load(String id, String word);

    CompletionStage<Map<String, Integer>> loadAll(String id);

    CompletionStage<Done> save(String id, String word, int count);
  }
  //#repository

  public class CassandraWordCountRepository implements WordCountRepository {
    private final CassandraSession session;

    final String keyspace = "test";
    final String table = "wordcount";
    private final String keyspaceTable = keyspace + "." + table;

    public CassandraWordCountRepository(CassandraSession session) {
      this.session = session;
    }

    @Override
    public CompletionStage<Integer> load(String id, String word) {
      return session.selectOne("SELECT count FROM " + keyspaceTable + " WHERE id = ? and word = ?", id, word)
          .thenApply(maybeRow -> {
            if (maybeRow.isPresent())
              return maybeRow.get().getInt("count");
            else
              return 0;
          });
    }

    @Override
    public CompletionStage<Map<String, Integer>> loadAll(String id) {
      return session.selectAll("SELECT word, count FROM " + keyspaceTable + " WHERE id = ?", id)
          .thenApply(rows -> {
            return rows
                .stream()
                .collect(Collectors.toMap(row -> row.getString("word"), row -> row.getInt("count")));
          });
    }

    @Override
    public CompletionStage<Done> save(String id, String word, int count) {
      return session.executeWrite(
          "INSERT INTO " + keyspaceTable + " (id, word, count) VALUES (?, ?, ?)",
          id,
          word,
          count);
    }

    public CompletionStage<Done> createKeyspaceAndTable() {
      return session
          .executeDDL(
              "CREATE KEYSPACE IF NOT EXISTS " + keyspace + " WITH REPLICATION = { 'class' : 'SimpleStrategy','replication_factor':1 }")
          .thenCompose(done ->
              session.executeDDL(
                  "CREATE TABLE IF NOT EXISTS " + keyspaceTable + " (\n" +
                      "  id text, \n" +
                      "  word text, \n" +
                      "  count int, \n" +
                      "  PRIMARY KEY (id, word)) \n"));
    }
  }

  //#sourceProvider
  class WordSource implements SourceProvider<Long, WordEnvelope> {

    private final Source<WordEnvelope, NotUsed> src = Source.from(
        Arrays.asList(new WordEnvelope(1L, "abc"), new WordEnvelope(2L, "def"), new WordEnvelope(3L, "ghi"), new WordEnvelope(4L, "abc")));

    @Override
    public CompletionStage<Source<WordEnvelope, ?>> source(Supplier<CompletionStage<Optional<Long>>> offset) {
      return offset.get().thenApply(o -> {
        if (o.isPresent())
          return src.dropWhile(envelope -> envelope.offset <= o.get());
        else
          return src;
      });
    }

    @Override
    public Long extractOffset(WordEnvelope envelope) {
      return envelope.offset;
    }

  }
  //#sourceProvider

  interface IllustrateVariables {
    //#mutableState
    public class WordCountHandler extends Handler<WordEnvelope> {
      private final Logger logger = LoggerFactory.getLogger(getClass());
      private final Map<String, Integer> state = new HashMap<>();


      @Override
      public CompletionStage<Done> process(WordEnvelope envelope) {
        String word = envelope.word;
        int newCount = state.getOrDefault(word, 0) + 1;
        logger.info("Word count for {} is {}", word, newCount);
        state.put(word, newCount);
        return CompletableFuture.completedFuture(Done.getInstance());
      }
    }
    //#mutableState
  }

  interface IllustrateStatefulHandlerLoadingInitialState {
    //#loadingInitialState
    public class WordCountHandler extends StatefulHandler<Map<String, Integer>, WordEnvelope> {
      private final ProjectionId projectionId;
      private final WordCountRepository repository;

      public WordCountHandler(ProjectionId projectionId, WordCountRepository repository) {
        this.projectionId = projectionId;
        this.repository = repository;
      }

      @Override
      public CompletionStage<Map<String, Integer>> initialState() {
        return repository.loadAll(projectionId.id());
      }

      @Override
      public CompletionStage<Map<String, Integer>> process(Map<String, Integer> state, WordEnvelope envelope) {
        String word = envelope.word;
        int newCount = state.getOrDefault(word, 0) + 1;
        CompletionStage<Map<String, Integer>> newState =
            repository
                .save(projectionId.id(), word, newCount)
                .thenApply(done -> {
                  state.put(word, newCount);
                  return state;
                });

        return newState;
      }
    }
    //#loadingInitialState
  }

  interface IllustrateStatefulHandlerLoadingStateOnDemand {
    //#loadingOnDemand
    public class WordCountHandler extends StatefulHandler<Map<String, Integer>, WordEnvelope> {
      private final ProjectionId projectionId;
      private final WordCountRepository repository;

      public WordCountHandler(ProjectionId projectionId, WordCountRepository repository) {
        this.projectionId = projectionId;
        this.repository = repository;
      }

      @Override
      public CompletionStage<Map<String, Integer>> initialState() {
        return CompletableFuture.completedFuture(new HashMap<>());
      }

      @Override
      public CompletionStage<Map<String, Integer>> process(Map<String, Integer> state, WordEnvelope envelope) {
        String word = envelope.word;

        CompletionStage<Integer> currentCount;
        if (state.containsKey(word))
          currentCount = CompletableFuture.completedFuture(state.get(word));
        else
          currentCount = repository.load(projectionId.id(), word);

        CompletionStage<Map<String, Integer>> newState =
            currentCount.thenCompose(n -> {
              return repository.save(projectionId.id(), word, n + 1)
                  .thenApply(done -> {
                    state.put(word, n + 1);
                    return state;
                  });
            });

        return newState;
      }
    }
    //#loadingOnDemand
  }


}
